File: android/app/src/main/java/com/rafaeltmbr/stopwatch/domain/use_cases/impl/PauseStopwatchUseCaseImpl.kt
package com.rafaeltmbr.stopwatch.domain.use_cases.impl

import com.rafaeltmbr.stopwatch.domain.entities.Status
import com.rafaeltmbr.stopwatch.domain.entities.StopwatchState
import com.rafaeltmbr.stopwatch.domain.services.TimerService
import com.rafaeltmbr.stopwatch.domain.data.stores.MutableStateStore
import com.rafaeltmbr.stopwatch.domain.use_cases.PauseStopwatchUseCase

class PauseStopwatchUseCaseImpl(
    private val store: MutableStateStore<StopwatchState>,
    private val timerService: TimerService
) : PauseStopwatchUseCase {
    override suspend fun execute() {
        if (!timerService.state.value.isRunning) return

        timerService.pause()
        store.update { it.copy(status = Status.PAUSED) }
    }
}

File: ios/Stopwatch/domain/use_cases/impl/PauseStopwatchUseCaseImpl.swift
class PauseStopwatchUseCaseImpl<MSS, TS>: PauseStopwatchUseCase
where MSS: MutableStateStore, MSS.State == StopwatchState, TS: TimerService {
    private let store: MSS
    private let timer: TS
    
    init(_ store: MSS, _ timer: TS) {
        self.store = store
        self.timer = timer
    }
    
    func execute() async {
        guard timer.state.isRunning else { return }
        
        timer.pause()
        await store.update { StopwatchState(status: .paused, milliseconds: $0.milliseconds) }
    }
}
